File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/booking/PackageBookingScreen.kt
package lk.mageride.passenger.booking

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Info
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.SegmentedButton
import androidx.compose.material3.SegmentedButtonDefaults
import androidx.compose.material3.SingleChoiceSegmentedButtonRow
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.passenger.R
import lk.mageride.passenger.onboarding.PhoneNumber
import lk.mageride.passenger.ui.MoneyFormat
import lk.mageride.passenger.ui.component.InlineError
import lk.mageride.passenger.ui.component.LabelledTextField
import lk.mageride.passenger.ui.component.MageRideCta
import lk.mageride.passenger.ui.component.PhoneNumberField
import lk.mageride.passenger.ui.component.SectionLabel
import lk.mageride.passenger.ui.theme.ControlTokens
import lk.mageride.passenger.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.PackageSize
import lk.mageride.shared.data.models.ride.RidePaymentMethod
import org.koin.androidx.compose.koinViewModel

/**
 * SCR-PA-012 — sending a parcel (US-20.1/20.2/20.8, P-06).
 *
 * The wireframe, top to bottom: `‹ Send a package`, the **S/M/L** selector with the ⓘ hint that
 * *"updates per pick"*, a description, the recipient's name and number, **Pickup location** with
 * three methods, **Drop-off location** with four, `Payment: COD ▾`, and *"Get estimate & Book"*.
 *
 * **The two ends offer different methods and that is the fence.** Pickup: Search / Map / Paste
 * link. Drop-off: those plus **Request**, which asks the *recipient* to share where the parcel
 * should go. There is nobody at the pickup to ask — the sender is standing there.
 *
 * The **Map** method is a modal on this screen for the reason [MapPickSheet] gives: the wireframe
 * assigns no SCR-PA id to a map picker, so there is no route to navigate to.
 */
@Composable
@Suppress("LongMethod") // The wireframe's form, in the wireframe's order.
internal fun PackageBookingScreen(
    onBack: () -> Unit,
    onSearch: (PackageEnd) -> Unit,
    onBooked: (String) -> Unit,
    model: PackageBookingViewModel = koinViewModel(),
) {
    val state by model.state.collectAsStateWithLifecycle()
    var pasteFor by remember { mutableStateOf<PackageEnd?>(null) }
    var mapFor by remember { mutableStateOf<PackageEnd?>(null) }

    LaunchedEffect(state.booked) {
        // The OTP is shown once and never again (P-07), so the screen surfaces it before the
        // navigation rather than after — see PackageBookingViewModel.
        state.booked?.takeIf { state.pickupOtp == null }?.let {
            onBooked(it)
            model.onBookingConsumed()
        }
    }

    Column(modifier = Modifier.fillMaxSize()) {
        Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            IconButton(onClick = onBack) {
                Icon(
                    imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                    contentDescription = stringResource(R.string.action_back),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            Text(
                text = stringResource(R.string.package_title),
                style = MaterialTheme.typography.titleLarge,
                color = MaterialTheme.colorScheme.onSurface,
            )
        }

        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = MageRideTheme.spacing.md),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
        ) {
            SectionLabel(text = stringResource(R.string.package_size))
            SizeSelector(size = state.size, onSelect = model::setSize)

            // P-06's hint. Directly below the size box and swapping with it, which is what
            // change619 #2 added and what stops a sender picking S for a fridge.
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
            ) {
                Icon(
                    imageVector = Icons.Filled.Info,
                    contentDescription = null,
                    modifier = Modifier.size(ControlTokens.ChipIcon),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                Text(
                    text = stringResource(state.sizeHint),
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            LabelledTextField(
                label = stringResource(R.string.package_description),
                value = state.description,
                onValueChange = model::onDescriptionChanged,
                placeholder = stringResource(R.string.package_description_hint),
                keyboardType = KeyboardType.Text,
                imeAction = ImeAction.Next,
            )

            LabelledTextField(
                label = stringResource(R.string.package_recipient_name),
                value = state.recipientName,
                onValueChange = model::onRecipientNameChanged,
                placeholder = stringResource(R.string.package_recipient_name_hint),
                keyboardType = KeyboardType.Text,
                imeAction = ImeAction.Next,
            )

            SectionLabel(text = stringResource(R.string.package_recipient_phone))
            PhoneNumberField(
                value = state.recipientPhone,
                onValueChange = model::onRecipientPhoneChanged,
                countryCode = PhoneNumber.COUNTRY_CODE,
                placeholder = PhoneNumber.PLACEHOLDER,
            )

            EndCapture(
                title = stringResource(R.string.package_pickup),
                end = PackageEnd.PICKUP,
                methods = PACKAGE_PICKUP_METHODS,
                state = state,
                model = model,
                onSearch = onSearch,
                onPickOnMap = { mapFor = it },
                onPaste = { pasteFor = it },
            )

            EndCapture(
                title = stringResource(R.string.package_dropoff),
                end = PackageEnd.DROPOFF,
                methods = ALL_METHODS,
                state = state,
                model = model,
                onSearch = onSearch,
                onPickOnMap = { mapFor = it },
                onPaste = { pasteFor = it },
            )

            SectionLabel(text = stringResource(R.string.package_payment))
            PaymentRow(method = state.paymentMethod, onChange = model::setPaymentMethod)

            state.estimateMinor?.let { amount ->
                Text(
                    text = stringResource(R.string.package_estimate, MoneyFormat.rupees(amount)),
                    style = MaterialTheme.typography.titleMedium,
                    color = MaterialTheme.colorScheme.onSurface,
                )
            }

            // P-07 — shown once, never returned again. The sender gives it to the driver.
            state.pickupOtp?.let { otp ->
                Text(
                    text = stringResource(R.string.package_pickup_otp, otp),
                    style = MaterialTheme.typography.titleMedium,
                    color = MageRideTheme.status.success,
                )
                MageRideCta(
                    label = stringResource(R.string.package_otp_noted),
                    onClick = {
                        state.booked?.let(onBooked)
                        model.onBookingConsumed()
                    },
                )
            }

            state.error?.let { InlineError(message = stringResource(it)) }

            if (state.pickupOtp == null) {
                MageRideCta(
                    label = if (state.estimateMinor == null) {
                        stringResource(R.string.package_get_estimate)
                    } else {
                        stringResource(R.string.package_book)
                    },
                    onClick = { if (state.estimateMinor == null) model.estimate() else model.book() },
                    enabled = state.canEstimate,
                    loading = state.estimating || state.booking,
                    modifier = Modifier.padding(bottom = MageRideTheme.spacing.md),
                )
            }
        }
    }

    pasteFor?.let { end ->
        PasteLinkSheet(
            label = stringResource(
                if (end == PackageEnd.PICKUP) R.string.paste_label_pickup else R.string.paste_label_dropoff,
            ),
            onUse = { place ->
                model.setPlace(end, place)
                pasteFor = null
            },
            onPickOnMap = {
                pasteFor = null
                mapFor = end
            },
            onDismiss = { pasteFor = null },
        )
    }

    mapFor?.let { end ->
        MapPickSheet(
            label = stringResource(
                if (end == PackageEnd.PICKUP) R.string.package_pickup else R.string.package_dropoff,
            ),
            around = (if (end == PackageEnd.PICKUP) state.pickup else state.dropoff)?.point,
            onUse = { place ->
                model.setPlace(end, place)
                mapFor = null
            },
            onDismiss = { mapFor = null },
        )
    }
}

/** P-06's `( S )( M )( L )`. */
@Composable
private fun SizeSelector(size: PackageSize, onSelect: (PackageSize) -> Unit) {
    SingleChoiceSegmentedButtonRow(modifier = Modifier.fillMaxWidth()) {
        PackageSize.entries.forEachIndexed { index, option ->
            SegmentedButton(
                selected = option == size,
                onClick = { onSelect(option) },
                shape = SegmentedButtonDefaults.itemShape(index = index, count = PackageSize.entries.size),
            ) {
                // S / M / L are the same character in all three scripts, so they are the enum's
                // own names rather than three identical `strings.xml` values.
                Text(option.name)
            }
        }
    }
}

/** One end's label, method row and captured value. */
@Composable
private fun EndCapture(
    title: String,
    end: PackageEnd,
    methods: List<PickupMethod>,
    state: PackageBookingState,
    model: PackageBookingViewModel,
    onSearch: (PackageEnd) -> Unit,
    onPickOnMap: (PackageEnd) -> Unit,
    onPaste: (PackageEnd) -> Unit,
) {
    SectionLabel(text = title)
    LocationMethodRow(
        methods = methods,
        selected = if (end == PackageEnd.PICKUP) state.pickupMethod else state.dropoffMethod,
        onSelect = { method ->
            model.setMethod(end, method)
            when (method) {
                PickupMethod.SEARCH -> onSearch(end)

                PickupMethod.MAP -> onPickOnMap(end)

                PickupMethod.PASTE_LINK -> onPaste(end)

                // The recipient is asked over the same P-02 round trip a proxy rider is. C080's
                // SCR-PA-020 owns the tracking that follows; the request itself is SCR-PA-010b's
                // machinery and is reached from there. See the C079 handoff.
                PickupMethod.REQUEST -> Unit
            }
        },
    )
    CapturedPlace(
        place = if (end == PackageEnd.PICKUP) state.pickup else state.dropoff,
        emptyLabel = stringResource(R.string.package_not_set),
    )
}

/** Cash / LankaQR / OnePay / **COD** — the last of which exists only on this screen (US-20.8). */
@Composable
private fun PaymentRow(method: RidePaymentMethod, onChange: (RidePaymentMethod) -> Unit) {
    SingleChoiceSegmentedButtonRow(modifier = Modifier.fillMaxWidth()) {
        RidePaymentMethod.entries.forEachIndexed { index, option ->
            SegmentedButton(
                selected = option == method,
                onClick = { onChange(option) },
                shape = SegmentedButtonDefaults.itemShape(index = index, count = RidePaymentMethod.entries.size),
            ) {
                Text(stringResource(paymentLabel(option)))
            }
        }
    }
}

File: apps/passenger-ios/PassengerApp/Booking/PackageBookingScreen.swift
import MageRideShared
import SwiftUI

/// SCR-PI-012 — a parcel instead of a person (US-20.1/20.2, P-06).
///
/// The cell: `‹ Back · Send a package`, a segmented **S / M / L** with the size hint under it, a
/// grouped list of description and recipient, a **Pickup location** picker with three methods and a
/// **Drop-off location** picker with four, the paste/request note, the COD payment row, and *"Get
/// estimate & Book"*.
///
/// **The two ends are not symmetric, and that is a fence rather than an omission.** The sender is
/// standing at the pickup, so there is nobody there to ask; the recipient is the only person who
/// knows where the parcel is going. ``PackageEnd`` carries which methods each end offers and
/// ``PackageBookingModel/setMethod(_:_:)`` refuses the fourth on a pickup even if a layout ever
/// offered it.
///
/// **The pickup OTP is shown once** (P-07) — see ``PackageBookingModel``.
@MainActor
struct PackageBookingScreen: View {

    @StateObject private var model: PackageBookingModel

    private let bookings: BookingRepository

    /// The Map method's geocoder and opening camera — see ``ProxyRiderScreen`` for both.
    private let places: PassengerPlaces
    private let lastFix: LastKnownFix

    let onBack: () -> Void
    let onSearch: (PackageEnd) -> Void
    let onBooked: (String) -> Void

    @State private var pasteEnd: PackageEnd?
    @State private var mapEnd: PackageEnd?

    init(
        draft: BookingDraft,
        bookings: BookingRepository,
        keys: IdempotencyKeys,
        otps: PackageOtps,
        places: PassengerPlaces,
        lastFix: LastKnownFix,
        onBack: @escaping () -> Void,
        onSearch: @escaping (PackageEnd) -> Void,
        onBooked: @escaping (String) -> Void
    ) {
        _model = StateObject(
            wrappedValue: PackageBookingModel(draft: draft, bookings: bookings, keys: keys, otps: otps)
        )
        self.bookings = bookings
        self.places = places
        self.lastFix = lastFix
        self.onBack = onBack
        self.onSearch = onSearch
        self.onBooked = onBooked
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
                SectionLabel(key: "package_size")
                Picker("", selection: sizeBinding) {
                    ForEach(Self.sizes, id: \.self) { size in
                        Text(Self.sizeLabel(size)).tag(size)
                    }
                }
                .pickerStyle(.segmented)
                .labelsHidden()
                .accessibilityLabel(Text(key: "package_size"))

                InfoBanner(messageKey: model.state.sizeHintKey, tone: .muted, symbolName: "info.circle")

                details

                SectionLabel(key: "package_pickup")
                LocationMethodPicker(
                    methods: PackageEnd.pickup.methods,
                    selection: model.state.pickupMethod,
                    onSelect: { choose(.pickup, $0) }
                )
                CapturedPlaceRow(place: model.state.pickup, emptyKey: "package_not_set")

                SectionLabel(key: "package_dropoff")
                LocationMethodPicker(
                    methods: PackageEnd.dropoff.methods,
                    selection: model.state.dropoffMethod,
                    onSelect: { choose(.dropoff, $0) }
                )
                CapturedPlaceRow(place: model.state.dropoff, emptyKey: "package_not_set")

                Text(key: "package_capture_note")
                    .mageFont(.caption)
                    .foregroundStyle(MageRideColor.onSurfaceVariant)

                // COD is booking-time and package-only (AL-22, US-20.8), which is why this list is
                // `PaymentRails.parcel` and not SCR-PI-009's three.
                PaymentChipRow(
                    method: model.state.paymentMethod,
                    rails: PaymentRails.parcel,
                    onChange: { model.setPaymentMethod($0) }
                )

                estimateRow

                if let errorKey = model.state.errorKey {
                    FormErrorText(messageKey: errorKey)
                }

                Button(action: submit) {
                    Text(key: model.state.estimateMinor == nil ? "package_get_estimate" : "package_book")
                }
                .buttonStyle(.mageCta(loading: model.state.isEstimating || model.state.isBooking))
                .disabled(!model.state.canEstimate)
            }
            .padding(MageRideSpacing.md)
        }
        .background(MageRideColor.surface)
        .navigationTitle(Text(key: "package_title"))
        .navigationBarTitleDisplayMode(.inline)
        .onAppear { model.refreshPlaces() }
        .onChange(of: model.state.booked) { rideId in
            guard let rideId else { return }
            onBooked(rideId)
            model.onBookingConsumed()
        }
        .sheet(item: $pasteEnd) { end in
            PasteLinkSheet(
                bookings: bookings,
                target: end == .pickup ? .pickup : .dropoff,
                onUse: { model.setPlace(end, $0) },
                onPickOnMap: { mapEnd = end },
                onDismiss: { pasteEnd = nil }
            )
        }
        .sheet(item: $mapEnd) { end in
            MapPickSheet(
                places: places,
                titleKey: end == .pickup ? "package_pickup" : "package_dropoff",
                around: (end == .pickup ? model.state.pickup : model.state.dropoff)?.point ?? lastFix.point,
                onUse: { model.setPlace(end, $0) },
                onDismiss: { mapEnd = nil }
            )
        }
    }

    private var details: some View {
        GroupedList {
            GroupedRow(titleKey: "package_description") {
                TextField(
                    "package_description_hint".localised,
                    text: Binding(get: { model.state.packageDescription }, set: { model.onDescriptionChanged($0) })
                )
                .mageFont(.body)
                .foregroundStyle(MageRideColor.onSurface)
                .multilineTextAlignment(.trailing)
            }
            GroupedRow(titleKey: "package_recipient_name") {
                TextField(
                    "package_recipient_name_hint".localised,
                    text: Binding(get: { model.state.recipientName }, set: { model.onRecipientNameChanged($0) })
                )
                .mageFont(.body)
                .foregroundStyle(MageRideColor.onSurface)
                .multilineTextAlignment(.trailing)
                .textContentType(.name)
            }
            GroupedRow(titleKey: "package_recipient_phone", showsSeparator: false) {
                TextField(
                    PhoneNumber.placeholder,
                    text: Binding(get: { model.state.recipientPhone }, set: { model.onRecipientPhoneChanged($0) })
                )
                .mageFont(.body)
                .foregroundStyle(MageRideColor.onSurface)
                .multilineTextAlignment(.trailing)
                .keyboardType(.phonePad)
                .textContentType(.telephoneNumber)
            }
        }
    }

    /// The quote, once there is one, and the *"same Mode C fare"* line the cell puts beside it
    /// (US-20.9).
    @ViewBuilder
    private var estimateRow: some View {
        if let amount = model.state.estimateMinor {
            HStack(spacing: MageRideSpacing.xs) {
                Text(MoneyFormat.rupees(amount))
                    .mageFont(.title)
                    .foregroundStyle(MageRideColor.onSurface)
                Text(key: "package_same_fare")
                    .mageFont(.caption)
                    .foregroundStyle(MageRideColor.onSurfaceVariant)
                Spacer(minLength: 0)
            }
        }
    }

    /// One CTA for two steps: a price has to exist before a booking can be made, because the token
    /// is what binds it. The label says which step it is on.
    private func submit() {
        if model.state.estimateMinor == nil {
            model.estimate()
        } else {
            model.book()
        }
    }

    private func choose(_ end: PackageEnd, _ method: PickupMethod) {
        model.setMethod(end, method)
        switch method {
        case .search: onSearch(end)
        case .map: mapEnd = end
        case .pasteLink: pasteEnd = end
        // The recipient round trip is the one method with no wired answer yet — see the C097
        // handoff, which carries C079's fifth gap forward unchanged.
        case .request: break
        }
    }

    private var sizeBinding: Binding<PackageSize> {
        Binding(get: { model.state.size }, set: { model.setSize($0) })
    }

    /// P-06's three, in the wireframe's order. Written out rather than read off `entries`, which is
    /// a Kotlin static this bridge does not promise a spelling for.
    private static let sizes: [PackageSize] = [PackageSize.s, PackageSize.m, PackageSize.l]

    /// `S` / `M` / `L` — a letter, and the same one in all three scripts, so it is a value rather
    /// than three identical `.strings` entries.
    ///
    /// Switched rather than read off `KotlinEnum.name`: the exported spelling of a Kotlin enum's
    /// `name` is the compiler's business, and three characters are not worth depending on it for.
    private static func sizeLabel(_ size: PackageSize) -> String {
        switch size {
        case PackageSize.m: return "M"
        case PackageSize.l: return "L"
        default: return "S"
        }
    }
}

extension PackageEnd: Identifiable {
    /// `sheet(item:)` needs one, and there are exactly two ends.
    var id: String { self == .pickup ? "pickup" : "dropoff" }
}
